plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,611 @@
+		if (!fu_common_read_uint16_safe (res, sizeof(res), 0x3,
+						 checksum, G_LITTLE_ENDIAN, error))
+			return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_fw_object_create (FuPxiDevice *self, const FuChunk *chk, GError **error)
+{
+	guint8 opcode = 0;
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+
+	/* request */
+	fu_byte_array_append_uint8 (req, PXI_HID_DEV_OTA_OUTPUT_REPORT_ID);
+	fu_byte_array_append_uint8 (req, FU_PXI_DEVICE_CMD_FW_OBJECT_CREATE);
+	fu_byte_array_append_uint32 (req, chk->address, G_LITTLE_ENDIAN);
+	fu_byte_array_append_uint32 (req, chk->data_sz, G_LITTLE_ENDIAN);
+	if (!fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0x0,
+					 req->data, req->len, error))
+		return FALSE;
+
+	/* check object create success or not */
+	if (!fu_pxi_device_wait_notify (self, 0x0, &opcode, NULL, error))
+		return FALSE;
+	if (opcode != FU_PXI_DEVICE_CMD_FW_OBJECT_CREATE) {
+		 g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_READ,
+			     "FwObjectCreate opcode got 0x%02x, expected 0x%02x",
+			     opcode,
+			     FU_PXI_DEVICE_CMD_FW_OBJECT_CREATE);
+		 return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_write_payload (FuPxiDevice *self, const FuChunk *chk, GError **error)
+{
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+	fu_byte_array_append_uint8 (req, PXI_HID_DEV_OTA_OUTPUT_REPORT_ID);
+	g_byte_array_append (req, chk->data, chk->data_sz);
+	return fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0x0,
+					   req->data, req->len, error);
+}
+
+static gboolean
+fu_pxi_device_write_chunk (FuPxiDevice *self, const FuChunk *chk, GError **error)
+{
+	guint32 prn = 0;
+	guint16 checksum = fu_pxi_device_calculate_checksum (chk->data, chk->data_sz);
+	guint16 checksum_device = 0;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* send create fw object command */
+	if (!fu_pxi_device_fw_object_create (self, chk, error))
+		return FALSE;
+
+	/* write payload */
+	chunks = fu_chunk_array_new (chk->data, chk->data_sz, chk->address,
+				     0x0, self->mtu_size);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk2 = g_ptr_array_index (chunks, i);
+		if (!fu_pxi_device_write_payload (self, chk2, error))
+			return FALSE;
+		prn++;
+		/* wait notify from device when PRN over threshold write or
+		 * offset reach max object sz or write offset reach fw length */
+		if (prn >= self->prn_threshold || i == chunks->len - 1) {
+			guint8 opcode = 0;
+			if (!fu_pxi_device_wait_notify (self, 0x0,
+							&opcode,
+							&checksum_device,
+							error))
+				return FALSE;
+			if (opcode != FU_PXI_DEVICE_CMD_FW_WRITE) {
+				g_set_error (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_READ,
+					     "FwWrite opcode invalid 0x%02x",
+					     opcode);
+				return FALSE;
+			}
+			prn = 0;
+		}
+	}
+
+	/* the last chunk */
+	self->checksum += checksum;
+	g_debug ("checksum 0x%04x, table checksum 0x%04x", checksum_device, self->checksum);
```

Comment:
since you're already comparing the two of these and setting an error when they fail, it seems not useful to include

Reply:
agree, removed.